src/main/scala/nl.knaw.dans.easy.solr4files/ApplicationWiring.scala

```diff
@@ -15,15 +15,88 @@
+import nl.knaw.dans.easy.solr4files.components._
+import nl.knaw.dans.lib.error.{ CompositeException, TraversableTryExtensions }
 import nl.knaw.dans.lib.logging.DebugEnhancedLogging
 
+import scala.util.{ Failure, Success, Try }
+
 /**
  * Initializes and wires together the components of this application.
  *
  * @param configuration the application configuration
  */
-class ApplicationWiring(configuration: Configuration) extends DebugEnhancedLogging
-// Mix in components
-{
-  // Configure components with configuration settings.
+class ApplicationWiring(configuration: Configuration)
+  extends DebugEnhancedLogging with VaultIO with Vault with Solr {
+
+  // don't need resolve for solr, URL gives more early errors TODO perhaps not enough early errors
+  override val solrUrl: URL = new URL(configuration.properties.getString("solr.url", ""))
+  override val vaultBaseUri: URI = new URI(configuration.properties.getString("vault.url", ""))
+
+  def initAllStores(): Try[FeedBackMessage] = {
+    getStoreNames
+      .flatMap(storeNames => updateStores(storeNames))
+      .map(results => s"Updated all bags of ${ results.size } stores ($vaultBaseUri)")
+  }
+
+  def initSingleStore(storeName: String): Try[FeedBackMessage] = {
+    getBagIds(storeName)
+      .flatMap(bagIds => updateBags(storeName, bagIds))
+      .map(results => s"Updated ${ results.size } bags of one store ($storeName)")
+  }
+
+  def update(storeName: String, bagId: String): Try[FeedBackMessage] = {
+    val bag = Bag(storeName, bagId, this)
+    for {
+      ddmXML <- bag.loadDDM
+      ddm = new DDM(ddmXML)
+      filesXML <- bag.loadFilesXML
+      files = (filesXML \ "file").map(FileItem(bag, ddm, _)).filter(f => !f.path.isEmpty && f.accessRights!="NONE")
+      _ <- deleteBag(bag.bagId)
+      feedbackMessage <- createDocs(bag, ddm, files)
+      _ <- commit()
+    } yield feedbackMessage
+  }
+
+  def delete(bagId: String): Try[FeedBackMessage] = for {
+    _ <- deleteBag(bagId)
+    _ <- commit()
+  } yield s"Deleted file documents for bag $bagId"
+
+  private def updateStores(storeNames: Seq[String]): Try[Seq[FeedBackMessage]] = {
+    storeNames
+      .map(initSingleStore)
+      .collectResults.recoverWith {
+      case t: CompositeException => throw new Exception(s"Tried to update ${ storeNames.size } stores, ${ t.getMessage() }", t)
+    }
+  }
+
+  private def updateBags(storeName: String, bagIds: Seq[String]): Try[Seq[FeedBackMessage]] = {
+    bagIds
+      .map(uuid => update(storeName, uuid))
+      .collectResults.recoverWith {
+      case t: CompositeException => throw new Exception(s"Tried to update ${ bagIds.size } bags, ${ t.getMessage() }", t)
+    }
+  }
+
+  private def createDocs(bag: Bag, ddm: DDM, files: Seq[FileItem]): Try[FeedBackMessage] = {
+    val results = files.map(fileItem => createDoc(bag, ddm, fileItem))
+    collectMixedResults(bag.bagId, results)
+  }
+
+  private def collectMixedResults(bagId: String, results: Seq[Try[FeedBackMessage]]) = {
```

Comment:
after reading this method more carefully, a couple of things come to mind:

1. why don't you inline this method with `createDocs`? It's only used there, after all...
2. I'd use a different implementation:
    a. first do a `results.collectResults`, which has type `Try[Seq[FeedBackMessage]]`
    b. if this results in a `Success`, you know that the number of `Failure`s in `results` is zero. The `successMessages` can then directly be found inside the `Success`
    c. if the result is a `Failure`, you get a `Failure(CompositeExceptions(listOfExceptions))`. You can either leave this as is, or wrap it in another `Exception` as you do now.

Reply:
1: I extracted it to stress the similarity with the normal collectResult. Moved it to an implicit class.
2c: still might have to collect some actions that did succeed and count (as in 2b) how many were indexed with or without content. But I could drop one loop.

Reply:
You may also be interested in a `seq.partition(...)` function?